src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs

```diff
@@ -1770,17 +1770,23 @@ internal static string ConcatenateStringPathArguments(CommandElementAst stringAs
             {
                 RemoveLastNullCompletionResult(result);
 
-                string enumString = LanguagePrimitives.EnumSingleTypeConverter.EnumValues(parameterType);
-                string separator = CultureInfo.CurrentUICulture.TextInfo.ListSeparator;
-                string[] enumArray = enumString.Split(separator, StringSplitOptions.RemoveEmptyEntries);
+                var enumValues = parameterType.GetEnumValues().Cast<Enum>().ToList();
```

Comment:
I think we have no need to do such conversions again and again since we already have the values in TypeConverter hash and on other hand ValidateElement does the conversion too.
So I suggest add new helper in EnumSingleTypeConverter class and use it here (and there is another one more place in CompleteStatementFlags() where we can use it):
```c#
            internal static string[] GetEnumValues(Type enumType)
                => EnumSingleTypeConverter.GetEnumHashEntry(enumType).names;
```

Reply:
They don't have the same requirements unfortunately. `ValidateElement()` will cast to a common `_promotedType`. For `[ValidateRange(1,3)][ConsoleColor]$color` or `[ValidateRange([ConsoleColor]::Black,10)][ConsoleColor]$color` that will be `Int32` and not the enum-type, so we need to provide the `Enum` directly for the validation to work.

I can add a `Array GetEnumValues(Type enumType)` returning `GetEnumHashEntry(enumType).values`. Though it feels a bit wrong to add methods unrelated to conversion to `EnumSingleTypeConverter`. Is it that much faster than `Enum.GetValues(parameterType)` directly?

I can submit a separate code cleanup-PR for `CompleteStatementFlags()` to use a  `internal string[] GetEnumNames(Type)` helper if you'd like.

Reply:
PowerShell is smart enough if we use right type:
```powershell
function ColorValidation {
   param(
      [ValidateRange([ConsoleColor]::Black, [ConsoleColor]::Gray)]
      [ConsoleColor]$color
   )
   Write-Output "Color validated!!!"
}

ColorValidation "Black"
ColorValidation "White"
```

Reply:
True, but it's a different scenario. In that demo, the arguments is already converted to enum-type before validation -> ValidateElement() can always compare ConsoleColor or int (converted if min/max is int or mixed type).

When calling it from the completer, we don't have that safeguard, so with mixed type min/max in the attribute, passing the string to `FilterElements()`->`ValidateElement()` would throw when trying to convert enum name string to `_promotedType` which is int in that case.

Anyhow, I've added a helper and cleanup up a bit. Not sure how to avoid the `.Cast<Enum>.ToList()` called once since the enum values are returned as `Array`. If possible, let me know 🙂 I appreciate the feedback so I may learn.

Reply:
> When calling it from the completer, we don't have that safeguard, so with mixed type min/max in the attribute, passing the string to `FilterElements()`->`ValidateElement()` would throw when trying to convert enum name string to `_promotedType` which is int in that case.

Int? I wonder. Do you mean `ColorValidation 0<Tab>`? I don't think we have to worry about this scenario.

Reply:
Yes. Attribute ctor converts to a common type, ex.

```powershell
[ValidateRange]::new(1, [System.ConsoleColor]::DarkCyan).MaxRange.GetType() # int32 even though MaxRange arg was enum
```

So if you call `ValidateElement("Black")` on the attribute above it would throw in a inner call to `TryConvertTo()` when trying to convert to "Black" string to int32 (the calculated common type set in `_promotedType`). 

If the attribute was created with enum-type for both Min-/MaxRange the call would work, but technically the code above is allowed.

Reply:
I don't understand why an user would specify int in the attribute. Enum is exactly what you need to use meaningful names instead of numbers.

Reply:
100% agree.

It is intentionally supported though so this PR should be compatible with it, which it is atm. :) Might be because of `[ValidateRange(rangeKind)]` so you could allow only positive values in some advanced scenarios